### PR TITLE
Change llm api return value to vineyard::Status.

### LINKED ISSUE #1791: llm-cache: Use `vineyard:Status` as function results
image

### PR DESCRIPTION
**Pull request by vegetableysm**
What do these changes do?
-------------------------

Related issue number
--------------------

Fixes #1791

### CODE REVIEW

File: modules/llm-cache/ds/kv_state_cache.cc [hunk 1]
```diff
@@ -115,8 +115,7 @@ KVStateCacheBuilder::KVStateCacheBuilder(Client& client,
 KVStateCacheBlockBuilder* KVStateCacheBuilder::Split(
     Client& client, KVStateCacheBlockBuilder* kvStateCacheBlockBuilder,
     std::vector<std::shared_ptr<NodeData>> nodeDataList) {
-  // Split the tree if the list of kvState is full.
-  VINEYARD_ASSERT(nodeDataList.size() > 0);
+  // Split the tree if the list of kvState is full.ß
```
Comment (dashanji) [inline modules/llm-cache/ds/kv_state_cache.cc:118]: Please delete the extra characters.

Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache.cc]: Done

File: modules/llm-cache/ds/kv_state_cache.cc [hunk 2]
```diff
@@ -106,20 +112,27 @@ KVStateCacheBuilder::KVStateCacheBuilder(Client& client,
...
     KVStateCacheBlockBuilder* kvStateCacheBlockBuilder =
         new KVStateCacheBlockBuilder(client, kvStateCacheBlock);
+    if (kvStateCacheBlockBuilder == nullptr) {
+      throw std::runtime_error("Not enough memory for new block builder.");
+    }
 
     treeData->kvStateCacheBlockBuilder = kvStateCacheBlockBuilder;
     treeData->isPtr = true;
   }
 }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache.cc:122]: DO NOT throw exceptions if you could return a status.

You could add a private constructor and add something like `Make()` instead, e.g.,

```
class KVStateCacheBlockBuilder {
public:
  Status Make(std::shared_ptr<KVStateCacheBlockBuilder> &out, ....);
}
```

File: modules/llm-cache/ds/kv_state_cache_block.cc [hunk 3]
```diff
@@ -129,8 +129,8 @@ KVStateCacheBlockBuilder::KVStateCacheBlockBuilder(
 }
 
 // current we do not consider the layer.
-int KVStateCacheBlockBuilder::Query(Client& client, int index,
-                                    KV_STATE_WITH_LAYER& kvState) {
+Status KVStateCacheBlockBuilder::Query(Client& client, int index,
+                                       KV_STATE_WITH_LAYER& kvState) {
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_block.cc:133]: You may check that the `index` won't out-of-range here.

File: modules/llm-cache/ds/kv_state_cache_block.cc [hunk 4]
```diff
@@ -164,16 +164,19 @@ bool KVStateCacheBlockBuilder::IsFull() {
...
     K_STATE keyState = (kvState.find(currentLayer)->second).first;
     V_STATE valueState = (kvState.find(currentLayer)->second).second;
-    VINEYARD_ASSERT(keyState.length ==
-                    (size_t) this->dimension * sizeof(double));
-    VINEYARD_ASSERT(valueState.length ==
-                    (size_t) this->dimension * sizeof(double));
+    if (keyState.length != (size_t) this->dimension * sizeof(double) ||
+        valueState.length != (size_t) this->dimension * sizeof(double)) {
+      return Status::Invalid("Update: invalid data.");
+    }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_block.cc:179]: Use `RETURN_ON_ASSERT(....)`

File: modules/llm-cache/radix-tree/radix-tree.cc [hunk 5]
```diff
@@ -490,12 +490,10 @@ std::shared_ptr<RadixTree> RadixTree::Deserialize(std::string data) {
 
     raxNode* node = nullptr;
     VLOG(100) << "stage 1";
-    VINEYARD_ASSERT(radixTree->tree != nullptr);
 
-    // TBD refator this code.
+    // TBD refactor this code.
     node = raxFindAndReturnDataNode(radixTree->tree, subTreeTokenList[i].data(),
                                     subTreeTokenList[i].size(), NULL, false);
-    VINEYARD_ASSERT(node != nullptr);
```
Comment (sighingnow) [inline modules/llm-cache/radix-tree/radix-tree.cc:498]: Use `RETURN_ON_ASSERT(..)` if you need to check some condtions.

File: modules/llm-cache/ds/kv_state_cache.cc (hunk 2 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache.cc:122]: Currently I add MakeKVStateCacheManager function to create KVStateCacheManager, add try...catch to catch exception throw by creating cache builder and block builder and return error status. How about this change?

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 6]
```diff
@@ -76,13 +76,31 @@ KVStateCacheManager::KVStateCacheManager(int dimension, int cacheCapacity,
...
+  if (syncThread == nullptr) {
+    throw std::runtime_error("Create sync thread failed!");
+  }
 
   // TBD
   // use lease to prevent the deadlock if the client is down
 }
 
+Status KVStateCacheManager::MakeKVStateCacheManager(
+    KVStateCacheManager*& manager, int dimension, int cacheCapacity, int layer,
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:89]: `MakeKVStateCacheManager` -> `Make`, as when you use the static method you would type `KVStateCacheManager::Make()`

File: modules/llm-cache/tests/kv_state_cache_test.cc [hunk 7]
```diff
@@ -44,8 +44,13 @@ KVStateCacheManager* kv_state_cache_manager;
...
           std::string socket) {
-  kv_state_cache_manager = new KVStateCacheManager(dimension, capacity, layer,
-                                                   block_size, 3, socket);
+  Status status = KVStateCacheManager::MakeKVStateCacheManager(
+      kv_state_cache_manager, dimension, capacity, layer, block_size, 3,
+      socket);
+  if (!status.ok()) {
+    LOG(ERROR) << "Failed to create KVStateCacheManager: " << status.ToString();
+    exit(1);
+  }
```
Comment (sighingnow) [inline modules/llm-cache/tests/kv_state_cache_test.cc:53]: Use `VINEYARD_CHECK_OK(KVStateCacheManager::Make(...))` in test cases.

File: modules/llm-cache/tests/kv_state_cache_benchmark_test.cc [hunk 8]
```diff
@@ -35,8 +35,13 @@ using namespace vineyard;  //  NOLINT(build/namespaces)
...
 
 void init() {
-  manager =
-      new KVStateCacheManager(DIMENSION, CAPACITY, LAYER, DEFAULT_BLOCK_SIZE);
+  Status status = KVStateCacheManager::MakeKVStateCacheManager(
+      manager, DIMENSION, CAPACITY, LAYER, BLOCK_SIZE, 3,
+      std::string(getenv("VINEYARD_IPC_SOCKET")));
+  if (!status.ok()) {
+    LOG(ERROR) << "Failed to create KVStateCacheManager: " << status.ToString();
+    exit(1);
```
Comment (sighingnow) [inline modules/llm-cache/tests/kv_state_cache_benchmark_test.cc:43]: Use `VINEYARD_CHECK_OK(KVStateCacheManager::Make(...))` in test cases.

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 9]
```diff
@@ -76,13 +76,31 @@ KVStateCacheManager::KVStateCacheManager(int dimension, int cacheCapacity,
...
+    int blockSize, int syncInterval, std::string socket) {
+  try {
+    manager = new KVStateCacheManager(dimension, cacheCapacity, layer,
+                                      blockSize, syncInterval, socket);
+    if (manager == nullptr) {
+      return Status::NotEnoughMemory("Failed to make KVStateCacheManager.");
+    }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:96]: I believe the `new` operator (without `-fno-expection`) never returns `nullptr`, rather, it throws `std::bad_alloc`.

**Review by sighingnow — CHANGES_REQUESTED** [review]
Remember to revise the constructors of:

- https://github.com/vegetableysm/v6d/blob/fix-return-value-of-llm-api/modules/llm-cache/ds/kv_state_cache.cc#L70-L122

File: modules/llm-cache/ds/kv_state_cache.cc (hunk 2 above)
Comment (sighingnow) [reply modules/llm-cache/ds/kv_state_cache.cc:122]: `MakeKVStateCacheManager` -> `Make()`, as my previous review comment.

> add try...catch to catch exception throw by creating cache builder and block builder and return error status.

Use the static `Status Make(...)` as well.

**Comment by sighingnow** [comment]
ONLY use try catch when there's a known exception thrown from the thirdparty libraries, otherwise, vineyard uses `Status` to propogate errors.

File: modules/llm-cache/ds/kv_state_cache_block.cc (hunk 3 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_block.cc:132]: OK

File: modules/llm-cache/ds/kv_state_cache_block.cc (hunk 4 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_block.cc]: OK

File: modules/llm-cache/radix-tree/radix-tree.cc (hunk 5 above)
Comment (vegetableysm) [reply modules/llm-cache/radix-tree/radix-tree.cc:498]: OK
Now I use RETURN_ON_ASSERT and RETURN_ON_ERROR to check condition.

File: modules/llm-cache/tests/kv_state_cache_test.cc (hunk 7 above)
Comment (vegetableysm) [reply modules/llm-cache/tests/kv_state_cache_test.cc]: OK

File: modules/llm-cache/tests/kv_state_cache_benchmark_test.cc (hunk 8 above)
Comment (vegetableysm) [reply modules/llm-cache/tests/kv_state_cache_benchmark_test.cc]: OK

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 9 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc]: Ok, I understand.

File: modules/llm-cache/ds/kv_state_cache_block.cc (hunk 3 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_block.cc:132]: OK

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 6 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc]: Done

**Review by dashanji — APPROVED** [review]

File: modules/llm-cache/ds/kv_state_cache.cc [hunk 10]
```diff
@@ -110,16 +117,24 @@ KVStateCacheBuilder::KVStateCacheBuilder(Client& client,
     treeData->kvStateCacheBlockBuilder = kvStateCacheBlockBuilder;
     treeData->isPtr = true;
   }
+
+  kvStateCacheBuilder =
+      std::shared_ptr<KVStateCacheBuilder>(new KVStateCacheBuilder(
+          client, cache->GetDimension(), cache->GetLayer(), cache->rootTree));
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache.cc:123]: No `std::shared_ptr<T>(new T(...))`, use `std::make_shared<T>(...)`.

Please double check all such occurences in llm-cache/ module.

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 11]
```diff
@@ -25,21 +25,32 @@ limitations under the License.
...
+    int syncInterval, std::string& llmCacheSyncLock,
+    std::string& llmCacheObjectName)
+    : client(client) {
   this->syncInterval = syncInterval;
-  VLOG(100) << "socket:" << socket;
-  client.Connect(socket);
+  this->kvStateCacheBuilder = cache;
+  this->llmCacheSyncLock = llmCacheSyncLock;
+  this->llmCacheObjectName = llmCacheObjectName;
+  this->syncThread = new std::thread(SyncThreadFunc, this);
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:37]: No `std::thread *` please. Why a pointer is needed here?

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 12]
```diff
@@ -50,73 +61,100 @@ KVStateCacheManager::KVStateCacheManager(int dimension, int cacheCapacity,
...
+  while (1) {
+    client.TryReleaseLock(actualKey, result);
+    if (!result) {
+      VLOG(100) << "Failed to release the lock, wait for next time.";
+      sleep(1);
+      continue;
+    } else {
+      break;
+    }
+  }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:94]:
```suggestion
  while (!client.TryReleaseLock(actualKey, result)) {
    VLOG(100) << "Failed to release the lock, wait for next time.";
    sleep(1);
  }
```

**Review by sighingnow — CHANGES_REQUESTED** [review]
Use `std::shared_ptr<T>` for returns object for `Make()`.

Use smart pointers in C++ whenever you can to avoid memory leak / dangling pointer bugs.

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 13]
```diff
@@ -252,8 +290,31 @@ void KVStateCacheManager::SyncThreadFunc(KVStateCacheManager* manager) {
...
+      std::string actualKey;
+      bool result;
+      while (1) {
+        manager->client.TryAcquireLock(manager->llmCacheSyncLock, result,
+                                       actualKey);
+        if (result == true) {
+          break;
+        }
+        sleep(1);
+      }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:302]:
```suggestion
      while (!manager->client.TryAcquireLock(manager->llmCacheSyncLock, result,
                                       actualKey).ok() || !result) {
        sleep(1);
      }
```

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 14]
```diff
@@ -252,8 +290,31 @@ void KVStateCacheManager::SyncThreadFunc(KVStateCacheManager* manager) {
...
+
+      Status status = manager->Sync();
+      if (!status.ok()) {
+        LOG(ERROR) << "Sync failed: " << status.ToString() << " Retry later.";
+      }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:307]: Just a log even when `Sync()` failed?

File: modules/llm-cache/ds/kv_state_cache_manager.cc [hunk 15]
```diff
@@ -252,8 +290,31 @@ void KVStateCacheManager::SyncThreadFunc(KVStateCacheManager* manager) {
...
+
+      while (1) {
+        manager->client.TryReleaseLock(actualKey, result);
+        if (result == true) {
+          break;
+        }
+        sleep(1);
+      }
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache_manager.cc:315]:
```suggestion
      while (!manager->client.TryReleaseLock(actualKey, result).ok() || !result) {
        sleep(1);
      }
```

File: modules/llm-cache/radix-tree/radix-tree.cc [hunk 16]
```diff
@@ -210,10 +209,6 @@ std::string RadixTree::Serialize() {
   // raxShow(this->tree);
   std::string serializedStr;
 
-  if (tokenList.size() != dataList.size()) {
-    throw std::runtime_error(
-        "The size of token list and data list is not equal");
-  }
```
Comment (sighingnow) [inline modules/llm-cache/radix-tree/radix-tree.cc:216]: Just sliently ignore the error??

**Review by sighingnow — CHANGES_REQUESTED** [review]
Double check all usage of `while(1)` to see if it can be simplied.

When `while(1)` is indeed required, use `while(true)` or `do {...} while (true)` instead.

File: modules/llm-cache/radix-tree/radix-tree.cc (hunk 16 above)
Comment (vegetableysm) [reply modules/llm-cache/radix-tree/radix-tree.cc:216]: It is not an error.
In serialize, If it push one element to tokenlist, it must also push a element to datalist.

File: modules/llm-cache/tests/kv_state_cache_test.cc [hunk 17]
```diff
@@ -41,11 +41,14 @@ std::vector<int> round_4_tokens = {1, 2, 3, 4, 5, 6};
...
+Client client;
 
 void init(int dimension, int capacity, int layer, int block_size,
           std::string socket) {
-  kv_state_cache_manager = new KVStateCacheManager(dimension, capacity, layer,
-                                                   block_size, 3, socket);
+  client.Connect(socket);
+  VINEYARD_CHECK_OK(KVStateCacheManager::Make(client, kv_state_cache_manager,
+                                              dimension, capacity, layer,
+                                              block_size, 3));
```
Comment (sighingnow) [inline modules/llm-cache/tests/kv_state_cache_test.cc:51]: Why not `VINEYARD_CHECK_OK` on `client.Connect()`? I promise you would get another warning ...

File: modules/llm-cache/ds/kv_state_cache.cc (hunk 10 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache.cc]: OK

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 11 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc]: OK

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 12 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc]: I refactored it.

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 13 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc]: OK

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 14 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc:299]: I add a simple function to deal with failure.

File: modules/llm-cache/ds/kv_state_cache_manager.cc (hunk 15 above)
Comment (vegetableysm) [reply modules/llm-cache/ds/kv_state_cache_manager.cc]: I refactored it.

File: modules/llm-cache/tests/kv_state_cache_test.cc (hunk 17 above)
Comment (vegetableysm) [reply modules/llm-cache/tests/kv_state_cache_test.cc:51]: Sorry, this problem was originally addressed in another pr. I fix it now.

File: modules/llm-cache/ds/kv_state_cache.cc [hunk 18]
```diff
@@ -70,40 +70,47 @@ void KVStateCache::Resolve() {
...
 
-  std::shared_ptr<NodeData> rootTreeHeader = this->rootTree->GetRootNode();
+  std::shared_ptr<NodeData> rootTreeHeader = rootTree->GetRootNode();
   rootTreeHeader->treeData->data = treeData;
   rootTreeHeader->treeData->dataLength = sizeof(TreeData);
-  this->rootTree->SetSubtreeData(treeData);
+  rootTree->SetSubtreeData(treeData);
+
+  kvStateCacheBuilder = std::shared_ptr<KVStateCacheBuilder>(
+      new KVStateCacheBuilder(client, dimension, layer, rootTree));
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache.cc:100]: Use `std::make_shared`.

File: modules/llm-cache/ds/kv_state_cache.cc [hunk 19]
```diff
@@ -242,14 +263,17 @@ void KVStateCacheBuilder::Delete(std::shared_ptr<NodeData> evictedNodeData) {
...
-    return;
+    return Status::OK();
   }
 
-  std::shared_ptr<KVStateCacheBuilder> globalCacheBuilder =
-      std::make_shared<KVStateCacheBuilder>(client, kvStateCache);
+  std::shared_ptr<KVStateCacheBuilder> globalCacheBuilder;
+  Status status =
+      KVStateCacheBuilder::Make(client, globalCacheBuilder, kvStateCache);
+  RETURN_ON_ERROR(status);
```
Comment (sighingnow) [inline modules/llm-cache/ds/kv_state_cache.cc:275]:
```suggestion
  RETURN_ON_ERROR(KVStateCacheBuilder::Make(client, globalCacheBuilder, kvStateCache));
```

**Review by sighingnow — APPROVED** [review]

**Pull request merged**